core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala

```diff
@@ -281,16 +309,21 @@ class GroupMetadataManager(brokerId: Int,
 
             val responseError = group synchronized {
               if (status.error == Errors.NONE) {
-                if (!group.is(Dead)) {
+                if (!group.is(Dead) && !isTxnOffsetCommit) {
                   filteredOffsetMetadata.foreach { case (topicPartition, offsetAndMetadata) =>
                     group.completePendingOffsetWrite(topicPartition, offsetAndMetadata)
                   }
                 }
                 Errors.NONE
               } else {
                 if (!group.is(Dead)) {
+                  if (!group.hasPendingOffsetCommitsFromProducer(producerId))
+                    removeProducerGroup(producerId, group.groupId)
```

Comment:
@guozhangwang Hmm.. I'm not sure I understand the case you are talking about. If we had received some previous append from the same producer as part of a transaction, then it would be reflected in the `hasPendingOffsetCommitsFromProducer` call. Can you clarify the case?